README.md
# @fengsi/nuxt-axios-cache

## Setup

1. Add `@fengsi/nuxt-axios-cache` dependency to your project

```bash
yarn add @fengsi/nuxt-axios-cache # or npm install @fengsi/nuxt-axios-cache
```

2. Add `@fengsi/nuxt-axios-cache` to the `modules` section of `nuxt.config.js`

```js
export default {

  modules: [
    // Simple usage
    '@fengsi/nuxt-axios-cache',

    // With options
    ['@fengsi/nuxt-axios-cache', { 
      /* module options */
      cacheNodeTtl: 60 * 30,
      cacheNodeLimit: 300,
      clearCacheRoutePath: '/_/nuxt-cache' // Clear cache route, request method is DELETE
    }]
  ]
}
```

3. plugins/api.js
```js
import axios from 'axios'
import cacheAdapterEnhancer from '@fengsi/nuxt-axios-cache/dist/cacheAdapterEnhancer'

export default ({ ssrContext }) => {
  let adapter = null

  if (process.server) {
    adapter = cacheAdapterEnhancer(axios.defaults.adapter, {
      defaultCache: ssrContext.$axCache,
    })
  } else {
    adapter = cacheAdapterEnhancer(axios.defaults.adapter, {
      cacheBrowserEnable: true
    })
  }

  if (adapter && process.env.NODE_ENV !== 'development') {
    axios..create({ adapter })
  }
}
```

### default options
```js
  /* module options */
  updateAgeOnGet = true
  cacheNodeTtl = 1800
  cacheNodeLimit = 300
  clearCacheRoutePath = '/_/nuxt-cache' // 清除缓存路由， 请求方法为 DELETE

  /* cacheAdapterEnhancer options */
  enabledByDefault = true
  cacheFlag = 'cache'
  project = 'default' // key 值前缀
  defaultCache = null
  cacheBrowserEnable = false // 设置为 true 下面的参数才有作用
  cacheBrowserTtl = 3600
  cacheBrowserSession = true // 会话缓存
```

### Related configuration

<a href="https://github.com/isaacs/node-lru-cache" target="_blank">lru-cache</a> <br/>
<a href="https://github.com/localForage/localForage" target="_blank">localForage</a>

CHANGELOG.md
# Changelog

## [2.1.0](https://www.github.com/fengsi-io/nuxt-axios-cache/compare/v2.0.0...v2.1.0) (2021-12-21)


### Features

* 添加 cacheBrowserSession 选项，启用会话缓存 ([38582ab](https://www.github.com/fengsi-io/nuxt-axios-cache/commit/38582ab69d656bfd8caa27564e2e5abfafdf3fbd))

## [2.0.0](https://www.github.com/fengsi-io/nuxt-axios-cache/compare/v1.1.4...v2.0.0) (2021-07-10)


### ⚠ BREAKING CHANGES

* 不兼容修改

### Code Refactoring

* 重构api ([09c8e5c](https://www.github.com/fengsi-io/nuxt-axios-cache/commit/09c8e5cf98c32a830ed6ef04bf93b4d1550797d6))

### [1.1.4](https://www.github.com/fengsi-io/nuxt-axios-cache/compare/v1.1.3...v1.1.4) (2021-06-21)


### Bug Fixes

* 修改变量名含义 ([30833da](https://www.github.com/fengsi-io/nuxt-axios-cache/commit/30833dae0b5b2f96b888e07464783c37a0658767))

### [1.1.3](https://www.github.com/fengsi-io/nuxt-axios-cache/compare/v1.1.2...v1.1.3) (2021-06-16)


### Bug Fixes

* 设置清除缓存的请求方法只为DELETE ([b94d386](https://www.github.com/fengsi-io/nuxt-axios-cache/commit/b94d38610c41ec48bda517f01ca1443b52a48b27))
* 设置清除缓存的请求方法只为PATCH ([3fd1fba](https://www.github.com/fengsi-io/nuxt-axios-cache/commit/3fd1fba1543c2539d9efd4e37d6f54be90c255d7))

### [1.1.2](https://www.github.com/fengsi-io/nuxt-axios-cache/compare/v1.1.1...v1.1.2) (2021-06-15)


### Bug Fixes

* delete regexp ([23a3f87](https://www.github.com/fengsi-io/nuxt-axios-cache/commit/23a3f874f288eb8bfe8ffcc02a4a22b04bd55bec))

### [1.1.1](https://www.github.com/fengsi-io/nuxt-axios-cache/compare/v1.1.0...v1.1.1) (2021-06-15)


### Bug Fixes

* ie regexp error ([bd11c49](https://www.github.com/fengsi-io/nuxt-axios-cache/commit/bd11c49b94572be66f7b53451322d243c66057fe))

## [1.1.0](https://www.github.com/fengsi-io/nuxt-axios-cache/compare/v1.0.0...v1.1.0) (2021-06-12)


### Features

* nuxt-axios-cache ([2b62b04](https://www.github.com/fengsi-io/nuxt-axios-cache/commit/2b62b04dd19f811fc9bf9d1705f256fd563f00fd))


### Bug Fixes

* 包含dist文件 ([80718e0](https://www.github.com/fengsi-io/nuxt-axios-cache/commit/80718e033af26ef82c0702d6b3e80c28d8662823))


lib/cacheAdapterEnhancer.js
import localForage from 'localforage'
import buildURL from 'axios/lib/helpers/buildURL'

const browserCacheDatekey = 'open_browser_cache_date'

const forageCache = {
  get: async (key) => {
    const data = await localForage.getItem(key)
    if (!data) { return null }

    const { expire, value } = data
    if (expire < Date.now()) {
      localForage.removeItem(key)
      return null
    }
    return value
  },

  set: (key, value, expire = false, callback = false) => {
    if (expire && typeof expire === 'number') {
      expire = Math.round(expire * 1000 + Date.now()) // seconds
      localStorage.setItem(browserCacheDatekey, expire)
    }
    return localForage.setItem(key, { value, expire }, expire && callback)
  },

  del: (key) => {
    localForage.removeItem(key)
  },

  clear: () => {
    localForage.clear()
  }
}

function buildSortedURL (...args) {
  const builtURL = buildURL(...args)
  const [urlPath, queryString] = builtURL.split('?')
  if (queryString) {
    const paramsPair = queryString.split('&')
    return `${urlPath}?${paramsPair.sort().join('&')}`
  }
  return builtURL
}

function jsRequestIdleCallback (cb) {
  if (window.requestIdleCallback) {
    return window.requestIdleCallback(cb)
  }

  return setTimeout(cb, 1)
}

function clearCacheBrowser (cacheBrowserSession, cacheBrowserTtl) {
  const oldDate = localStorage.getItem(browserCacheDatekey)

  if (cacheBrowserSession || (oldDate && oldDate < Date.now())) {
    jsRequestIdleCallback(() => {
      forageCache.clear()
    })
  }

  const expire = Math.round(cacheBrowserTtl * 1000 + Date.now())
  localStorage.setItem(browserCacheDatekey, expire)
}

export default function cacheAdapterEnhancer (adapter, options = {}) {
  const {
    enabledByDefault = true,
    cacheFlag = 'cache',
    project = 'default',
    defaultCache = null,
    cacheBrowserTtl = 3600,
    cacheBrowserEnable = false,
    cacheBrowserSession = true
  } = options

  if (cacheBrowserEnable && window && window.localStorage) {
    // Clear client cache
    clearCacheBrowser(cacheBrowserSession, cacheBrowserTtl)
  }

  return async (config) => {
    const useCache = config[cacheFlag] ?? enabledByDefault
    const {
      url,
      method,
      params,
      paramsSerializer,
      forceUpdate,
      expireInSeconds = cacheBrowserTtl
    } = config

    if (method === 'get' && useCache) {
      const key = project + ':' + buildSortedURL(url, params, paramsSerializer)
      let responsePromise = null

      if (defaultCache) {
        responsePromise = defaultCache.get(key)

        if (!responsePromise || forceUpdate) {
          responsePromise = adapter(config)
            .then((response) => {
              defaultCache.set(key, response)
              return response
            })
            .catch((reason) => {
              defaultCache.del(key)
              throw reason
            })
        }

        return responsePromise
      }

      if (cacheBrowserEnable) {
        const item = await forageCache.get(key)

        if (!item || forceUpdate) {
          responsePromise = adapter(config)
            .then((response) => {
              const value = {
                data: response.data,
                status: response.status,
                config: null,
                request: null
              }
              forageCache.set(key, value, expireInSeconds, function (err) {
                if (err) {
                  console.log(err)
                }
              })
              return response
            })
            .catch((reason) => {
              forageCache.del(key)
              throw reason
            })
        } else {
          responsePromise = Promise.resolve(item)
        }

        return responsePromise
      }
    }

    return adapter(config)
  }
}


lib/module.js
import LRU from 'lru-cache'

module.exports = function (moduleOptions) {
  const {
    updateAgeOnGet = true,
    cacheNodeTtl = 1800,
    cacheNodeLimit = 300,
    clearCacheRoutePath = '/_/nuxt-cache'
  } = moduleOptions

  const axCache = new LRU({
    updateAgeOnGet,
    maxAge: Number(cacheNodeTtl) * 1000,
    max: Number(cacheNodeLimit)
  })

  this.nuxt.hook('vue-renderer:ssr:prepareContext', (ssrContext) => {
    ssrContext.$axCache = axCache
  })

  this.addServerMiddleware({
    path: clearCacheRoutePath,
    handler (req, res, next) {
      try {
        if (req.method === 'DELETE') {
          axCache.reset()
          res.setHeader('Content-Type', 'text/plain')
          res.end('success')
        } else {
          res.statusCode = 403
          res.statusMessage = 'Not found'
          res.end()
        }
      } catch (err) {
        next(err)
      }
    }
  })
}

module.exports.meta = require('../package.json')
